core/lib/instrument.c

```diff
@@ -1791,7 +1791,7 @@ create_and_initialize_module_data(app_pc start, app_pc end, app_pc entry_point,
             copy->segments[i].end = os_segments[i].end;
             copy->segments[i].prot = os_segments[i].prot;
         }
-    } else
+    } else if (segments != NULL)
```

Comment:
Probably the best thing to do would be to use an "assert" that the compiler can make use of when optimising, something like:
```
#ifdef DEBUG
# define ASSERT(x) ...
#else
# define ASSERT(x) do { if (!(x)) __builtin_unreachable(); } while (0)
#endif
```
So what exactly do you want? ASSERT, (redundant) check, and comment, or is the ASSERT plus (redundant) check enough?